controllers/openstackcluster_controller.go

```diff
@@ -682,3 +672,51 @@ func handleUpdateOSCError(openstackCluster *infrav1.OpenStackCluster, message er
 	openstackCluster.Status.FailureReason = &err
 	openstackCluster.Status.FailureMessage = pointer.String(message.Error())
 }
+
+// filterSubnets retrieves the subnets based on the Subnet filters specified on OpenstackCluster.
+func filterSubnets(networkingService *networking.Service, openStackCluster *infrav1.OpenStackCluster) ([]infrav1.Subnet, error) {
+	var subnets []infrav1.Subnet
+	openStackClusterSubnets := openStackCluster.Spec.Subnets
+	if len(openStackClusterSubnets) == 0 {
+		empty := &infrav1.SubnetFilter{}
+		listOpt := empty.ToListOpt()
+		listOpt.NetworkID = openStackCluster.Status.Network.ID
+		filteredSubnets, err := networkingService.GetSubnetsByFilter(listOpt)
+		if err != nil {
+			err = fmt.Errorf("failed to find subnets: %w", err)
+			if errors.Is(err, networking.ErrFilterMatch) {
+				handleUpdateOSCError(openStackCluster, err)
+			}
+			return nil, err
+		}
+		if len(filteredSubnets) > 2 {
+			return nil, fmt.Errorf("more than two subnets found in the Network. Specify the subnets in the OpenStackCluster.Spec instead")
+		}
+		for _, subnet := range filteredSubnets {
+			subnets = append(subnets, infrav1.Subnet{
```

Comment:
please write a function (in ports.go by preference and with unit tests please) to convert a filtered subnet into a infrav1.Subnet, and use it here and L713)

Reply:
I created the function and is tested indirectly tested by the "should allow two subnets for the cluster network" test scenario